transformation/src/main/scala/org/broadinstitute/monster/hca/HcaPipelineBuilder.scala

```diff
@@ -281,21 +287,31 @@ object HcaPipelineBuilder extends PipelineBuilder[Args] {
     isFileMetadata: Boolean = false
   ): ClosedTap[String] = {
     // get the readable files for the given input path
-    val readableFiles = getReadableFiles(
-      s"$inputPrefix/metadata/${entityType}/**.json",
-      context
-    )
-    // then convert json to msg and get the filename
-    val processedData = jsonToFilenameAndMsg(entityType)(readableFiles)
-      .withName(s"Pre-process ${entityType} metadata")
-      .map {
-        case (filename, metadata) =>
-          if (isFileMetadata) transformFileMetadata(entityType, filename, metadata)
-          else transformMetadata(entityType, filename, metadata)
-      }
+    val metadataFiles = getReadableFiles(s"$inputPrefix/metadata/${entityType}/**.json", context)
+    val metadataFilenameAndMsg = jsonToFilenameAndMsg(entityType)(metadataFiles)
+
+    // for file metadata
+    val readyMetadata = if (isFileMetadata) {
+      val descriptorFiles = getReadableFiles(
+        s"$inputPrefix/descriptors/$entityType/**.json",
+        context
+      )
+      val descriptorFilenameAndMsg = jsonToFilenameAndMsg(entityType)(descriptorFiles)
+      metadataFilenameAndMsg.join(descriptorFilenameAndMsg)
+    } else {
+      // for non-file metadata
+      jsonToFilenameAndMsg(entityType)(metadataFiles)
+    }
+
+    val processedMetadata = readyMetadata.withName(s"Pre-process $entityType metadata").map {
+      case (filename: String, metadata: Msg) => transformMetadata(entityType, filename, metadata)
+      case (filename: String, (metadata: Msg, descriptor: Msg)) =>
+        transformFileMetadata(entityType, filename, metadata, descriptor)
```

Comment:
Since these are totally-divergent paths, would prefer to push these transformation calls into the `if` / `else` above.